extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java

```diff
@@ -172,73 +161,91 @@ public void handle(RoutingContext event) {
+                    response.end();
+                    return;
+                }
             }
-
-            if (!allowsOrigin) {
-                LOG.debug("Origin is not allowed");
-                response.setStatusCode(403);
-                response.setStatusMessage("CORS Rejected - Invalid origin");
-                response.end();
-            } else if (!allowsMethod) {
+            if (allowedHeaders != null) {
+                response.headers().add(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS, allowedHeaders);
+            }
+            if (allowedMethods != null) {
+                response.headers().add(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, allowedMethods);
+            }
+            //we check that the actual request matches the allowed methods and headers
+            if (!isMethodAllowed(request.method())) {
                 LOG.debug("Method is not allowed");
                 response.setStatusCode(403);
                 response.setStatusMessage("CORS Rejected - Invalid method");
                 response.end();
-            } else if (request.method().equals(HttpMethod.OPTIONS) && (requestedHeaders != null || requestedMethods != null)) {
-                LOG.debug("Preflight request has completed");
-                if (corsConfig.accessControlMaxAge.isPresent()) {
-                    response.putHeader(HttpHeaders.ACCESS_CONTROL_MAX_AGE,
-                            String.valueOf(corsConfig.accessControlMaxAge.get().getSeconds()));
-                }
+                return;
+            }
+            if (!allowsOrigin) {
                 response.end();
-            } else {
-                event.next();
+                return;
             }
+
+            //all good, it can proceed
+            event.next();
         }
     }
 
+    private void handlePreflightRequest(RoutingContext event, String requestedHeaders, String requestedMethods, String origin,
+            boolean allowsOrigin) {
+        //see https://fetch.spec.whatwg.org/#http-cors-protocol
+
+        if (corsConfig.accessControlMaxAge.isPresent()) {
+            event.response().putHeader(HttpHeaders.ACCESS_CONTROL_MAX_AGE,
+                    String.valueOf(corsConfig.accessControlMaxAge.get().getSeconds()));
+        }
+        var response = event.response();
+        if (requestedMethods != null) {
+            processPreFlightMethods(response, requestedMethods);
+        }
+
+        if (requestedHeaders != null) {
+            processPreFlightRequestedHeaders(response, requestedHeaders);
+        }
+
+        //normal request
+        //always set expose headers if present
+        if (exposedHeaders != null) {
+            response.headers().add(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, exposedHeaders);
+        }
+
+        if (!isConfiguredWithWildcard(corsConfig.exposedHeaders)) {
+            response.headers().set(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, this.exposedHeaders);
+        }
```

Comment:
I don't fully understand what's going here, first we are adding a header value and then we are setting it? I think these blocks need more code comments

Reply:
Good catch, looks like a copy and paste problem, `exposedHeaders` will not be null only if it is not a wildcard config